TGEN: Added implementation of OMPL framework

pemami4911 opened the pull request:
Sub8StateSpace is an extension of OMPL's CompoundStateSpace.
The sub's state space is defined as a 12D vector. Currently,
the time derivative of the euler angles is used to be
the representation of the angular velocities,
and a unit quaternion is used for the orientation.

Planning will be carried out in the sub's control space;
accordingly, Sub8SpaceInformation is an extension of OMPL's
control::SpaceInformation.

StateValidityChecker allows the client to customize
collision checking, clearance from nearest invalid states,
etc.

Most unit tests are currently failing, since tests have
been written for unimplemented functions for Sub specific
behavior.

catkin_make was verified to build successfully.

pemami4911 committed:
TGEN: Added implementation of OMPL framework

Sub8StateSpace is an extension of OMPL's CompoundStateSpace.
The sub's state space is defined as a 12D vector. Currently,
the time derivative of the euler angles is used to be
the representation of the angular velocities,
and a unit quaternion is used for the orientation.

Planning will be carried out in the sub's control space;
accordingly, Sub8SpaceInformation is an extension of OMPL's
control::SpaceInformation.

StateValidityChecker allows the client to customize
collision checking, clearance from nearest invalid states,
etc.

Most unit tests are currently failing, since tests have
been written for unimplemented functions for Sub specific
behavior.

catkin_make was verified to build successfully.

jpanikulam commented on gnc/sub8_trajectory_generator/src/sub8_state_space.cc:
```diff
@@ -0,0 +1,34 @@
+#include "sub8_state_space.h"
+
+ompl::base::State* sub8::trajectory_generator::Sub8StateSpace::allocState() const {
+	StateType *state = new StateType();
```
shared_ptr possible?

forrestv replied:
Why are you using time derivatives of Euler angles for angular velocity
instead of just the angular velocity vector? From your comments, it looks
like that's what you're actually doing already ("Get the angular velocity
about the y axis").

At least from what I've seen, all that does is introduce more complex
Jacobians everywhere and add singularities. It was only used in the old
trajectory generator because it worked directly on each Euler angle
separately, but AFAIK your approach won't.

On Wed, Sep 23, 2015 at 3:47 PM, Jake notifications@github.com wrote:

> In gnc/sub8_trajectory_generator/src/sub8_state_space.cc
> https://github.com/uf-mil/Sub8/pull/20#discussion_r40248547:
> 
> > @@ -0,0 +1,34 @@
> > +/**
> > - \* Author: Patrick Emami
> > - \* Date: 9/21/15
> > - _/
> >   +
> >   +#include "sub8_state_space.h"
> >   +
> >   +ompl::base::State_ sub8::trajectory_generator::Sub8StateSpace::allocState() const {
> > -   StateType *state = new StateType();
> 
> shared_ptr possible?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/uf-mil/Sub8/pull/20/files#r40248547.

pemami4911 replied:
Right, so this was still a point of confusion for me at the time. I'll just be using the angular velocity vector, I see now how using the time derivative of the Euler angles wouldn't go over very well in this situation. I'll make the relevant changes to reflect this. 

> On Sep 23, 2015, at 3:57 PM, Forrest Voight notifications@github.com wrote:
> 
> In gnc/sub8_trajectory_generator/src/sub8_state_space.cc:
> 
> > @@ -0,0 +1,34 @@
> > +/**
> > - \* Author: Patrick Emami
> > - \* Date: 9/21/15
> > - _/
> >   +
> >   +#include "sub8_state_space.h"
> >   +
> >   +ompl::base::State_ sub8::trajectory_generator::Sub8StateSpace::allocState() const {
> > -   StateType *state = new StateType();
> >   Why are you using time derivatives of Euler angles for angular velocity instead of just the angular velocity vector? From your comments, it looks like that's what you're actually doing already ("Get the angular velocity about the y axis"). At least from what I've seen, all that does is introduce more complex Jacobians everywhere and add singularities. It was only used in the old trajectory generator because it worked directly on each Euler angle separately, but AFAIK your approach won't.
> >   …
> >   —
> >   Reply to this email directly or view it on GitHub.

jpanikulam commented on gnc/sub8_trajectory_generator/include/sub8_trajectory_generator/sub8_state_space.h:
```diff
@@ -0,0 +1,182 @@
+			// Get the maximum value a call to distance() can return (or an upper bound).
+			// For unbounded state spaces, this function can return infinity.
+			//
+    		// Tight upper bounds are preferred because the value of the extent is used
```
Sudden tab change?

pemami4911 committed:
TGEN: Addressing comments in PR#20

pemami4911 commented:
- The virtual function I'm overriding has a return value of ompl::base::State*, so using a raw pointer there is required and OK
- Changed a few function names in Sub8StateSpace to make the angular velocity getters more accurate, and fixed the tabulation problem (Sublime wasn't showing that there was spacing issues, but gedit did ?)

jpanikulam commented:
Re: spacing
Is it from mixing tabs and spaces?
- Jake

> On Sep 24, 2015, at 1:09 PM, Patrick E. notifications@github.com wrote:
> 
> The virtual function I'm overriding has a return value of ompl::base::State*, so using a raw pointer there is required and OK
> 
> Changed a few function names in Sub8StateSpace to make the angular velocity getters more accurate, and fixed the tabulation problem (Sublime wasn't showing that there was spacing issues, but gedit did ?)
> 
> —
> Reply to this email directly or view it on GitHub.

pemami4911 commented:
Yep haha 

> On Sep 24, 2015, at 1:44 PM, Jake notifications@github.com wrote:
> 
> Re: spacing
> Is it from mixing tabs and spaces?
> - Jake
> 
> > On Sep 24, 2015, at 1:09 PM, Patrick E. notifications@github.com wrote:
> > 
> > The virtual function I'm overriding has a return value of ompl::base::State*, so using a raw pointer there is required and OK
> > 
> > Changed a few function names in Sub8StateSpace to make the angular velocity getters more accurate, and fixed the tabulation problem (Sublime wasn't showing that there was spacing issues, but gedit did ?)
> > 
> > —
> > Reply to this email directly or view it on GitHub.
> > 
> > —
> > Reply to this email directly or view it on GitHub.

jpanikulam commented:
It is done
